src/null-safety/migrating-to-null-safety/index.md

```diff
@@ -0,0 +1,435 @@
+targets a version of Dart before null safety, you get the old behavior and your
+existing code keeps working. If your program targets the later null safe version
+of Dart, you opt in to the new type system and features.
+
+It's sort of like how the monolithic GCC compiler can compile code written in a
+number of different languages. It figures out what language each file is written
+in and compiles it appropriately.
+
+### Language version numbers
+
+Dart language versions are identified by a major and minor number that match the
+first two components of the Dart SDK. So, for example, the latest language
+version supported by Dart SDK 2.7.3 is Dart 2.7. We promise that each Dart SDK
+supports *all* of the language versions covered by its major version number. So
+Dart SDK 2.7.3 supports language 2.7, 2.6, 2.5, etc. Until we ship Dart 3.0.0,
+every SDK will continue to support legacy code that does *not* use null safety.
+
+Since the language version is derived from the SDK version, it implies:
+
+*   Whenever we ship a minor version of the SDK, a new language version appears.
+    In practice, many of these language versions are very similar to and
+    entirely compatible with previous versions. For example, Dart 2.9 (the
+    language) is essentially identical to 2.8.
+
+*   When we ship a patch version of the SDK, it cannot contain any backwards
+    incompatible language changes. Since a patch SDK release (like 2.7.2) does
+    not introduce a new language version (2.7) over the previous SDK (2.7.1,
+    also language 2.7), there would be no way to avoid breaking user code if
+    we did.
+
+### Specifying language version
+
+Every Dart SDK from **TODO: version** and later supports null safety. But they
+also all still support older null unsafe "legacy" code. To keep that code
+running, the SDK just needs to know that the code is targeting an older Dart
+language version. How do you tell the compiler that?
+
+We wanted to make the process of pinning your code to a language version as
+lightweight as possible. We wouldn't want to, say, use a different file
+extension like `.dartx` and make you rename every single one of your files every
+time a new language version came out. Instead, we build on top of what the tools
+already know about the package containing your code.
+
+Your [pubspec] probably already has an [SDK constraint] like:
+
+[pubspec]: https://dart.dev/tools/pub/pubspec
+[sdk constraint]: https://dart.dev/tools/pub/pubspec#sdk-constraints
+
+```dart
+environment:
+  sdk: '>=2.6.0 <3.0.0'
+```
+
+Here, your package claims that it can be run using any Dart SDK from 2.5.0 or
+later, until 3.0.0. For that to be true, your package can't be using any
+features that were introduced after 2.6.0. If you used, say, extension methods
+(released in 2.7.0), then your package wouldn't work if someone tried to run it
+on a Dart 2.6.0 SDK, which your own pubspec claims is supported.
```

Comment:
Is this right for the example above?

Reply:
Apologies, but I found this confusing. To me, ` sdk: '>=2.6.0 <3.0.0'` reads like I can use any SDK feature available between `[2.6.0, 3.0.0)`, but instead it means  SDK features available `<=2.6.0`.

Reply:
Yes, but if something states that it works with any SDK that is at least `2.6.0`, then `2.6.0` is one such SDK that it must work with, and hence it cant depend on a feature that wasn't added until `2.7.0`. 

Think of SDK constraints as listing all the versions you work with.